File: qabelbox/src/test/java/de/qabel/qabelbox/account/AccountManagerTest.java
package de.qabel.qabelbox.account;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import org.apache.commons.io.FileUtils;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.util.concurrent.Callable;

import de.qabel.qabelbox.BuildConfig;
import de.qabel.qabelbox.QabelBoxApplication;
import de.qabel.qabelbox.QblBroadcastConstants;
import de.qabel.qabelbox.SimpleApplication;
import de.qabel.qabelbox.TestConstants;
import de.qabel.qabelbox.config.AppPreference;
import de.qabel.qabelbox.helper.Formatter;
import de.qabel.qabelbox.storage.model.BoxQuota;
import de.qabel.qabelbox.storage.server.MockBlockServer;
import de.qabel.qabelbox.util.BoxTestHelper;
import de.qabel.qabelbox.util.TestHelper;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;

@RunWith(RobolectricGradleTestRunner.class)
@Config(application = SimpleApplication.class, constants = BuildConfig.class)
public class AccountManagerTest {

    private AccountManager accountManager;
    private AppPreference appPreferences;

    @Before
    public void setUp() {
        BoxTestHelper testHelper = new BoxTestHelper((QabelBoxApplication) RuntimeEnvironment.application);
        accountManager = testHelper.getAccountManager();
        appPreferences = testHelper.getAppPreferences();
        appPreferences.setAccountName("testUser");
        appPreferences.setAccountEMail("test@user.de");
        appPreferences.setToken(TestConstants.TOKEN);
        appPreferences.setBoxQuota(null);
    }

    @Test
    public void testLogout() {
        accountManager.logout();
        assertNull(appPreferences.getToken());
        assertNotNull(appPreferences.getAccountName());
        assertNotNull(appPreferences.getAccountEMail());
    }


    @Test
    public void testRefreshQuota() throws Exception {
        final int[] statusCodes = new int[]{0};
        BroadcastReceiver receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                int statusCode = intent.getIntExtra(QblBroadcastConstants.STATUS_CODE_PARAM, 0);
                statusCodes[0] = statusCode;
            }
        };
        RuntimeEnvironment.application.registerReceiver(receiver,
                new IntentFilter(QblBroadcastConstants.Account.ACCOUNT_CHANGED));

        BoxQuota quota = accountManager.getBoxQuota();

        assertEquals(0, quota.getSize());
        assertEquals(-1, quota.getQuota());

        TestHelper.waitUntil(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return statusCodes[0] > 0;
            }
        }, "Error waiting for quota");
        assertEquals(AccountStatusCodes.QUOTA_UPDATED, statusCodes[0]);
        quota = accountManager.getBoxQuota();

        assertEquals(MockBlockServer.SIZE, quota.getSize());
        assertEquals(MockBlockServer.QUOTA, quota.getQuota());

        //100Gb
        long oldQuota = MockBlockServer.QUOTA;
        long oldSize = MockBlockServer.SIZE;
        MockBlockServer.QUOTA = 107374182400L;
        MockBlockServer.SIZE = 1024L;
        statusCodes[0] = 0;

        quota = accountManager.getBoxQuota();

        assertEquals(oldSize, quota.getSize());
        assertEquals(oldQuota, quota.getQuota());
        TestHelper.waitUntil(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return statusCodes[0] > 0;
            }
        }, "Error waiting for changed quota");
        assertEquals(AccountStatusCodes.QUOTA_UPDATED, statusCodes[0]);
        quota = accountManager.getBoxQuota();

        assertEquals(MockBlockServer.SIZE, quota.getSize());
        assertEquals(MockBlockServer.QUOTA, quota.getQuota());

        assertEquals(FileUtils.byteCountToDisplaySize(quota.getQuota()), "100 GB");
        assertEquals(FileUtils.byteCountToDisplaySize(quota.getSize()), "1 KB");

    }
}

File: qabelbox/src/test/java/de/qabel/qabelbox/account/AccountManagerTest.kt
package de.qabel.qabelbox.account

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import de.qabel.qabelbox.*
import de.qabel.qabelbox.config.AppPreference
import de.qabel.qabelbox.storage.model.BoxQuota
import de.qabel.qabelbox.storage.server.MockBlockServer
import de.qabel.qabelbox.util.BoxTestHelper
import junit.framework.Assert.*
import org.apache.commons.io.FileUtils
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricGradleTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config

@RunWith(RobolectricGradleTestRunner::class)
@Config(application = SimpleApplication::class, constants = BuildConfig::class)
class AccountManagerTest {

    private lateinit var accountManager: AccountManager
    private lateinit var appPreferences: AppPreference

    private val GB_100: Long = 107374182400L
    private val KB_1: Long = 1024L

    var statusCodes = -1
    val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            statusCodes = intent.getIntExtra(QblBroadcastConstants.STATUS_CODE_PARAM, 0)
        }
    }

    @Before
    fun setUp() {
        val testHelper = BoxTestHelper(RuntimeEnvironment.application as QabelBoxApplication)
        accountManager = testHelper.accountManager
        appPreferences = testHelper.appPreferences.apply {
            accountName = "testUser"
            accountEMail = "test@user.de"
            token = TestConstants.TOKEN
            boxQuota = null
        }

        statusCodes = -1
        RuntimeEnvironment.application.registerReceiver(receiver,
                IntentFilter(QblBroadcastConstants.Account.ACCOUNT_CHANGED))
    }

    @Test
    fun testLogout() {
        accountManager.logout()
        with(appPreferences) {
            assertNull(token)
            assertNotNull(accountName)
            assertNotNull(accountEMail)
        }
    }

    @Test
    fun testRefreshQuota() {
        var quota = accountManager.boxQuota
        //Initial values
        quota.size eq 0
        quota.quota eq -1

        waitFor({ statusCodes > 0 }, "Error waiting for quota response!")

        assertEquals(AccountStatusCodes.QUOTA_UPDATED, statusCodes)
        quota = accountManager.boxQuota

        MockBlockServer.SIZE eq quota.size
        MockBlockServer.QUOTA eq quota.quota
    }

    @Test
    fun testRefreshUpdatedQuota() {
        val boxQuota = BoxQuota().apply {
            size = MockBlockServer.SIZE
            quota = MockBlockServer.QUOTA
        }
        appPreferences.boxQuota = boxQuota

        //100Gb
        MockBlockServer.QUOTA = GB_100
        MockBlockServer.SIZE = KB_1

        //Old values
        val outDatedQuota = accountManager.boxQuota
        boxQuota eq outDatedQuota

        waitFor({ statusCodes > 0 }, "Error waiting for changed quota response!")

        val updatedQuota = accountManager.boxQuota

        KB_1 eq updatedQuota.size
        GB_100 eq updatedQuota.quota

        "100 GB" eq FileUtils.byteCountToDisplaySize(updatedQuota.quota)
        "1 KB" eq FileUtils.byteCountToDisplaySize(updatedQuota.size)
    }

    fun waitFor(check: () -> Boolean, message: String, timeout: Long = 200) {
        val maximumTime = System.currentTimeMillis() + timeout
        val pollInterval = 100L
        while (System.currentTimeMillis() < maximumTime) {
            if (check()) {
                return
            }
            Thread.sleep(pollInterval)
        }
        fail(message)
    }
}
